[ADR] Choix des navigateurs web supportés (#5346)
1024pix/pix: docs/adr/0038-choix-des-navigateurs-supportes.md

> # Support navigateur

Reviewer: ```suggestion
# Support des navigateurs web
```

Date : 2022-12-29

## État

En cours

## Contexte

Pix est un service public, sans condition d'inscription. Il est en ligne, donc accessible via un navigateur web. Pix n'a pas la possibilité d'imposer un navigateur à ses utilisateurs, et doit pouvoir répondre à un large public, avec des navigateurs desktop/mobiles, obsolètes/à jour.

Nos frameworks et nos packages permettent de supporter les anciens navigateurs jusqu'à une certaine version.

> 

Reviewer: Il serait bon d'inclure une référence à la problématique écologique soulevée par @yannbertrand :

```suggestion
Le choix de supporter tels ou tels navigateurs a un impact écologique qu'il faut mentionner. D'un côté supporter des vieux navigateurs permet aux utilisateurs de pouvoir utiliser de vieux ordinateurs et téléphones et ainsi aux utilisateurs de garder leurs matériels informatiques plus longtemps. Mais d'un autre côté s'appuyer sur les nouvelles fonctionnalités natives des navigateurs récents (que ce soit au niveau de JavaScript ou des CSS) permet une réduction significative de la taille des paquets (bundles) produits téléchargés par les matériels des utilisateurs.

```

# Support navigateur

Date : 2022-12-??

## État

En cours

## Contexte

Pix est un service public, sans condition d'inscription. Il est en ligne, donc accessible via un navigateur web. Pix n'a pas la possibilité d'imposer un navigateur à ses utilisateurs, et doit pouvoir répondre à un large public, avec des navigateurs desktop/mobiles, obsolètes/à jour.

Nos frameworks et nos packages permettent de supporter les anciens navigateurs jusqu'à une certaine version.

Enfin, le support utilisateur a besoin de connaitre le plus facilement possible les navigateurs supportés par Pix.

> A noter que pour la certification, le cahier des charges des centres de certifications indiquent que les navigateurs doivent avoir moins de 2 ans, et que les établissements peuvent mettre à jour durant la période des vacances scolaires. Donc nous devons gérer au moins les navigateurs qui ont environ 3 ans.

Reviewer 2: Je propose de remplacer la dernière phrase

```suggestion
A noter que pour la certification, le cahier des charges des centres de certifications indiquent que les navigateurs doivent avoir moins de 2 ans, et que les établissements peuvent mettre à jour durant la période des vacances scolaires. Par sécurité, nous ajoutons un an à ce support minimum, nous gérons donc officiellement les navigateurs de moins de 3 ans.
```

Le support navigateur se fait via le fichier `/config/target.js` disponible dans les différents projets front. C'est Babel qui s'occupe de la rétrocompatibilité, en utilisant [BrowserList](https://github.com/browserslist/browserslist) et [Can I Use](https://caniuse.com/).

Ces sites utilisent les statistiques d'utilisations des navigateurs dans le monde.

Pour supporter les navigateurs très utilisés sur nos publics, anciens mais aussi nouveaux, il faut  mettre à jour la configuration du framework. 

De plus, [EmberJs ne supporte pas tous les navigateurs](https://emberjs.com/browser-support/) : Ember 4.0.0 ne supporte plus Internet Explorer, et globalement les navigateurs qui détiennent au moins 0,25 % de l'utilisation de la part de marché mondiale sur les mobiles et les ordinateurs de bureau, selon statcounter. 

## Décisions

Pour être au clair sur les navigateurs supportés, le fichier `/config/target.js`  doit indiquer :
- la version minimale à supporter pour les navigateurs les plus communs :
  - Chrome
  - Firefox
  - Safari
  - Opéra

>   - Edge

Reviewer 2: Je pense que cette liste peut être amenée à évoluer. Faut-il préciser ici cette liste ?

Author: Plus bas, on indique une liste "A date (Janvier 2023)," à date justement. Mais oui si de nouveaux navigateurs arrivent ou si on change la façon de choisir les navigateurs, il faudra refaire une ADR pour modifier celle-ci

Reviewer 2: Je propose de supprimer la liste ici et de préciser qu'elle est visible plus bas (ou sur une page spécifique si on choisit cette option)

> - les navigateurs utilisés à 1% dans le monde.

Reviewer 2: Est-ce vraiment une metric qui nous intéresse ? Je m'explique dans un autre commentaire

Author: Cela permet de se laisser un peu de marge par rapport à des navigateurs qu'on ne précise pas.
Vu l'état actuel du web, la liste précise de certains navigateurs ci-dessus englobe déjà les 1% les plus utilisés. Mais si un nouveau navigateur prend rapidement une grosse part de marché, nous pourrons le prendre en considération automatiquement.

Par conséquent, la liste des navigateurs est du type :
```
"browserslist": [
  "> 1%",
  "Firefox >= 52",
  "Chrome >= 55",
  "Edge >= 88",
  "Safari >= 88",

>   "Opera >= 88",

Reviewer 2: La [documentation de browserslist](https://github.com/browserslist/browserslist#full-list) précise qu'il est possible de préciser nos contraintes de manière plus explicite : 
- Pourcentage d'utilisation par pays. Ex : `> 1% in FR`
  * Il est même possible de fournir notre liste custom d'utilisation de navigateurs (`> 5% in my stats`)
- Durée depuis la sortie. Ex : `last 3 years`, `since 2020`, `since 2020-01-01`
- La version de Firefox toujours supporté uniquement : `Firefox ESR`
- ...

Ces règles peuvent [être composées](https://github.com/browserslist/browserslist#query-composition) de manière inclusive, exclusive ou rejetées avec des opérateurs : 
- `and` pour l'intersection des règles
- `or` pour l'union (règle utilisée si on sépare avec une virgule `,`)
- `not` pour l'exclusion

La liste de navigateurs peut être sortie à date avec la commande [`$ npx browserslist`](https://github.com/browserslist/browserslist#debug) si l'on souhaite l'afficher quelque part.

Ainsi, la liste des navigateurs supportées évoluerait d'elle-même. Je pense qu'on y gagnerait en clarté sur notre position et ça nous éviterait des revues régulières.

Reviewer: Je trouvais déjà cet ADR très bon 😃

J'ajoute :+1: pour décrire une formule browserlist complètement programmatique (mais incluant tout de même en dur la contrainte `"Firefox >= 52"` tant qu'elle est d'actualité) et que nous n'aurons donc pas à maintenir.

Author: Cet ADR propose d'utiliser la notation `"Firefox >= 52"` pour justement nous apporté de la visibilité sur les navigateurs vraiment supportés.
Lorsqu'on utilise `> 1% in FR`, il est difficile ensuite, pour le support utilisateur, de dire quelle version de Firefox nous supportons par exemple.
De même, `since 2020` est plus parlant, mais les personnes ne savent surement pas à quelle version cela fait référence.
Actuellement, la précision de la version semble la plus claire pour nous, même si elle doit effectivement évoluer avec le temps.

Reviewer: Après relecture de tout l'ensemble et la précision de @MelanieMEB qu'une contrainte forte est que l'info des navigateurs supportés doit être facilement trouvable et compréhensible, j'apporte mon soutien à la stratégie de lister explicitement les versions supportées+1% des navigateurs utilisés dans le monde.

Reviewer: Après relecture de tout l'ensemble et la précision de @MelanieMEB qu'une contrainte forte est que l'info des navigateurs supportés doit être facilement trouvable et compréhensible, j'apporte mon soutien à la stratégie de lister explicitement les versions supportées+1% des navigateurs utilisés dans le monde, même si la stratégie proposée par @yannbertrand est plus élégante d'un point de vue informatique.

Reviewer: Enfin, je trouve un peu perturbant pour la compréhension que la _« liste type des navigateurs supportés »_ indique des versions plus récentes que les versions données dans la _« liste des navigateurs supportés à date (Janvier 2023) »_. Pour que cela corresponde à ce à quoi le lecteur s'attend il faudrait que les versions de cette liste type soient inférieures ou égales aux versions de la liste à date de janvier 2023.

Reviewer 2: Merci pour vos retours :)

Une petite précision : on peut générer une page de manière programmatique contenant la liste des navigateurs supportés grâce à [`$ npx browserlist`](https://github.com/browserslist/browserslist#debug), je trouverai ça plus accessible pour le support et pour les utilisateurs souhaitant avoir cette info. (c'est compatible avec toutes les techniques de conditionnement).

Si on est prêt à intégrer le développement du paragraphe précédant, je propose donc cette alternative :

```
"browserslist": [
  "> 1%",
	"last 4 years",
  "Firefox >= 52",
]
```

Cette approche "non spécifique" a l'avantage de la pérennité, modulo Firefox qu'il faudra reprendre le jour où ce ne sera plus une contrainte pour Pix.

## Décisions

Pour être au clair sur les navigateurs supportés, le fichier `/config/target.js`  doit indiquer :
- la version minimale à supporter pour les navigateurs les plus communs :
  - Chrome
  - Firefox
  - Safari
  - Opéra
  - Edge
- les navigateurs utilisés à 1% dans le monde.

Par conséquent, la liste des navigateurs est du type :
```
"browserslist": [
  "> 1%",
  "Firefox >= 58",
  "Chrome >= 64",
  "Edge >= 41",
  "Safari >= 11",
  "Opera >= 50",
]
```

Les versions des navigateurs supportés sont celles qui ont moins de 4 ans à date pour laisser de la marge par rapport à la certification. 

Les versions minimales supportées seront mis à jour deux fois par an (fin d'année et l'été).

A date (Janvier 2023), cela signifie que le  `/config/target.js`  doit indiquer :
```
"browserslist": [
  ">= 1%",
  "Firefox >= 58",
  "Chrome >= 64",
  "Edge >= 41",
  "Safari >= 11",
  "Opera >= 50",
]
```

> Les versions supportées sont les mêmes pour toutes les applications Pix : les applications du repository Pix, Pix-UI et Pix Editor.

Reviewer: Préciser explicitement si cet ADR s'applique également aux pix-sites (https://pix.org/, https://pix.fr/, https://pro.pix.org/, https://pro.pix.fr/). Je suis pour que cet ADR  s'applique également aux pix-sites, de manière à ce que le parcours des utilisateurs soit complètement cohérent et qu'il ne pèse pas plus ou moins de contraintes entre les applications Pix et les pix-sites.
